### PR TITLE
graph: make edges and lines reversible

### PR DESCRIPTION
**Pull request by kortschak**
I'm sending this as a candidate implementation addressing #825. After having sat on this for a while, I'm fairly convinced this is the right thing to do; if fixes unfortunate subtleties and an otherwise insoluble bug (~~#417~~#824) at the cost of requiring an extra method on edge and line types.

I am not convinced of my approach to naming the methods because, as it stands (the two versions here), either a type assertion is needed for lines, or edges and lines are not interchangeable without wrapping. I think that probably the method names should be `ReversedEdge` and `ReversedLine` to avoid this collision.

Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
I'm going to leave this until @mewmew gives an opinion.

**Comment by mewmew** [comment]
> I'm going to leave this until @mewmew gives an opinion.

I think my comment is similar to the one I added before. It is unfortunate that we have to extend the number of methods of one of the core interfaces of the graph library. But if it solves a problem that cannot be solved (reasonably well) in another way, then, I'm not going to oppose it.

Still wish we could keep core interfaces to their very minimal, but this may be the purist part of me speaking, and not the pragmatic part.

**Comment by kortschak** [comment]
Yes, I understand the reluctance. Everyother strategy I have thought of is either more brittle or requires more differential method addition (we could have undirected and directed edges, but this massively complicates the API).

We are keeping things as simple as possible. But no simpler. Remember, generic graphs are hard.

I'll make the changes to the method names and then this can be reviewed.

**Comment by mewmew** [comment]
> Remember, generic graphs are hard.

I am starting to get an appreciation for this.

**Comment by kortschak** [comment]
I have not put a great deal of thought into that. However, I'm not convinced that it would impact on this issue one way or another.

**Comment by kortschak** [comment]
PTAL

The focus for review would probably be documentation changes and correctness of implementation additions (I found an error in the `WeightedEdgePair` method that would cause the returned edge to not be a weighted edge. (Maybe there should be a note in the reverse methods' documentation that it MUST return the same type as the receiver).

**Comment by kortschak** [comment]
*ping*

File: graph/testgraph/testgraph.go [hunk 1]
```diff
@@ -48,6 +48,27 @@ func checkEmptyIterator(t *testing.T, it graph.Iterator, useEmpty bool) {
 	}
 }
 
+// Edge is supports the simples edge operations.
```
Comment (vladimir-ch) [inline graph/testgraph/testgraph.go]: Delete "is", and is it "simple" or "simplest"?

Comment (kortschak) [reply graph/testgraph/testgraph.go]: `// Edge supports basic edge operations.`

**Comment by kortschak** [comment]
PTAL

**Comment by kortschak** [comment]
*ping*

**Pull request merged**